// car-query/dist/BodyStyle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BodyStyle = void 0;
var BodyStyle;
(function (BodyStyle) {
    BodyStyle["CargoVans"] = "Cargo Vans";
    BodyStyle["CompactCars"] = "Compact Cars";
    BodyStyle["Convertible"] = "Convertible";
    BodyStyle["Coupe"] = "Coupe";
    BodyStyle["Crossover"] = "Crossover";
    BodyStyle["Hatchback"] = "Hatchback";
    BodyStyle["LargeCars"] = "Large Cars";
    BodyStyle["MidsizeCars"] = "Midsize Cars";
    BodyStyle["MidsizeStationWagons"] = "Midsize Station Wagons";
    BodyStyle["MiniCompactCars"] = "Mini Compact Cars";
    BodyStyle["MiniVan"] = "Minivan";
    BodyStyle["model_body"] = "model_body";
    BodyStyle["NotAvailable"] = "Not Available";
    BodyStyle["PanelVan"] = "Panel Van";
    BodyStyle["PassengerVans"] = "Passenger Vans";
    BodyStyle["Pickup"] = "Pickup";
    BodyStyle["Roadster"] = "Roadster";
    BodyStyle["Sedan"] = "Sedan";
    BodyStyle["SmallPickupTrucks"] = "Small Pickup Trucks";
    BodyStyle["SmallSportUtilityVehicles"] = "Small Sport Utility Vehicles";
    BodyStyle["SmallStationWagons"] = "Small Station Wagons";
    BodyStyle["SportUtilityVehicles"] = "Sport Utility Vehicles";
    BodyStyle["StandardPickupTrucks"] = "Standard Pickup Trucks";
    BodyStyle["StandardSportUtilityVehicles"] = "Standard Sport Utility Vehicles";
    BodyStyle["StationWagon"] = "Station Wagon";
    BodyStyle["SubcompactCars"] = "Subcompact Cars";
    BodyStyle["SUV"] = "SUV";
    BodyStyle["TwoSeaters"] = "Two Seaters";
    BodyStyle["Van"] = "Van";
})(BodyStyle = exports.BodyStyle || (exports.BodyStyle = {}));
